x/nexus/keeper/transfer.go

```diff
@@ -60,8 +64,55 @@ func (k Keeper) getTransferFee(ctx sdk.Context) (fee exported.TransferFee) {
 	return fee
 }
 
+// computeChainSurcharge computes the surcharge fee over the base fee
+//
+// max_surcharge = chain.max_fee - chain.min_fee
+//
+// surcharge = min(max_surcharge, chain.fee_rate * (amount - baseFee))
+func computeChainSurcharge(feeInfo exported.FeeInfo, baseFee sdk.Int, amount sdk.Int) sdk.Int {
+	maxSurcharge := sdk.Int(feeInfo.MaxFee.Sub(feeInfo.MinFee))
+
+	remaining := sdk.ZeroInt()
+	if baseFee.LT(amount) {
+		remaining = amount.Sub(baseFee)
+	}
+
+	surcharge := sdk.NewDecFromInt(remaining).Mul(feeInfo.FeeRate).TruncateInt()
+	surcharge = sdk.MinInt(maxSurcharge, surcharge)
+
+	return surcharge
+}
+
+// ComputeTransferFee computes the fee for a cross-chain transfer.
+// If fee_info is not set for an asset on a chain, default of zero is used
+//
+// base_fee = source_chain.min_fee + destination_chain.min_fee
+//
+// transfer_fee = baseFee + source_chain_surcharge + destination_chain_surcharge
+//
+// INVARIANT: source_chain.min_fee + destination_chain.min_fee <= transfer_fee <= source_chain.max_fee + destination_chain.max_fee
+func (k Keeper) ComputeTransferFee(ctx sdk.Context, sourceChain exported.Chain, destinationChain exported.Chain, asset sdk.Coin) (sdk.Coin, exported.FeeInfo) {
+	sourceChainFeeInfo, _ := k.GetFeeInfo(ctx, sourceChain, asset.Denom)
+	destinationChainFeeInfo, _ := k.GetFeeInfo(ctx, destinationChain, asset.Denom)
+
+	baseFee := sdk.Int(sourceChainFeeInfo.MinFee.Add(destinationChainFeeInfo.MinFee))
```

Comment:
overflow check?

Reply:
We do rely on `Add` panicking due to overflow in confirm deposit. Might be easiest to do that here. Not sure what's a good way to check that without accessing the internal `BigInt` and making it more complicated.

Reply:
I wasn't concerned about the `.Add` between `sdk.UInt` but the conversion to `sdk.Int` in the end.